java/java.lsp.server/src/org/netbeans/modules/java/lsp/server/protocol/ProjectConfigurationCompletion.java

```diff
@@ -0,0 +1,122 @@
+ *
+ * @author Martin Entlicher
+ */
+@ServiceProvider(service = LaunchConfigurationCompletion.class, position = 100)
+public class ProjectConfigurationCompletion implements LaunchConfigurationCompletion {
+
+    private static final String CONFIG_TYPE = "java8+";     // NOI18N
+
+    @Override
+    public CompletableFuture<List<CompletionItem>> configurations(Supplier<CompletableFuture<Project>> projectSupplier) {
+        return projectSupplier.get().thenApply(p -> createConfigurationsCompletion(p));
+    }
+
+    @Override
+    public CompletableFuture<List<CompletionItem>> attributes(Supplier<CompletableFuture<Project>> projectSupplier, Map<String, Object> currentAttributes) {
+        return CompletableFuture.completedFuture(Collections.emptyList());
+    }
+
+    @Override
+    public CompletableFuture<List<CompletionItem>> attributeValues(Supplier<CompletableFuture<Project>> projectSupplier, Map<String, Object> currentAttributes, String attribute) {
+        if ("launchConfiguration".equals(attribute)) {      // NOI18N
+            return projectSupplier.get().thenApply(p -> createLaunchConfigCompletion(p));
+        } else {
+            return CompletableFuture.completedFuture(Collections.emptyList());
+        }
+    }
+
+    @NbBundle.Messages({"# {0} - Configuration name", "LBL_LaunchJavaConfig=Launch Java: {0}",
+                        "# {0} - Configuration name", "LBL_LaunchJavaConfig_desc=Launch a Java 8+ application using {0}."})
+    private static List<CompletionItem> createConfigurationsCompletion(Project p) {
+        ProjectConfigurationProvider<ProjectConfiguration> provider = p.getLookup().lookup(ProjectConfigurationProvider.class);
+        Collection<ProjectConfiguration> configurations = provider.getConfigurations();
+        List<CompletionItem> completionItems = new ArrayList<>(configurations.size() - 1);
+        boolean skipFirst = true;
+        for (ProjectConfiguration c : configurations) {
+            if (skipFirst) {
+                skipFirst = false;
+                continue;
+            }
+            String configDisplayName = c.getDisplayName();
+            String launchName = Bundle.LBL_LaunchJavaConfig(configDisplayName);
+            CompletionItem ci = new CompletionItem("Java 8+: " + launchName);   // NOI18N
+            StringWriter sw = new StringWriter();
+            try (JsonWriter w = new JsonWriter(sw)) {
+                w.setIndent("\t");                                          // NOI18N
+                w.beginObject();
+                w.name("name").value(launchName);                           // NOI18N
+                w.name("type").value(CONFIG_TYPE);                          // NOI18N
+                w.name("request").value("launch");                          // NOI18N
+                w.name("mainClass").value("${file}");                       // NOI18N
+                w.name("launchConfiguration").value(configDisplayName);     // NOI18N
```

Comment:
I wonder if we shouldn't finally add nonlocalizable IDs to the configuration API.

Reply:
I didn't seen that in other extensions. Did you?

Reply:
Sorry - that was for our NetBeans Project API. ProjectConfiguration API interface does not export its ID, but maven + gradle internally have one, to serve as persistent handle; nonlocalizable (while the displayName is localized).

Reply:
I don't think this is a problem. The user definitely does not want to have two configurations localized to the identical String. And when they are different, we don't have a problem.